GoRogue/MapGeneration/PolygonArea.cs

```diff
@@ -0,0 +1,174 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using SadRogue.Primitives;
+
+namespace GoRogue.MapGeneration
+{
+    /// <summary>
+    /// An area with an arbitrary number of sides and corners
+    /// </summary>
+    public class PolygonArea : IReadOnlyArea
+    {
+
+        /// <summary>
+        /// The corners of this polygon
+        /// </summary>
+        public IReadOnlyList<Point> Corners => _corners.AsReadOnly();
+        private readonly List<Point> _corners;
+
+        /// <summary>
+        /// The exterior points of the polygon
+        /// </summary>
+        public IReadOnlyArea OuterPoints => _outerPoints;
+        private readonly MultiArea _outerPoints;
+
+        /// <summary>
+        /// The interior points of the polygon
+        /// </summary>
+        public IReadOnlyArea InnerPoints => _innerPoints;
+        private readonly Area _innerPoints;
+
+        private readonly MultiArea _points;
+
+        /// <summary>
+        /// Which Line-Drawing algorithm to use
+        /// </summary>
+        public readonly Lines.Algorithm LineAlgorithm;
+
+        /// <summary>
+        /// Creates a new Polygon, with corners at the provided points
+        /// </summary>
+        /// <param name="corners">Each corner of the polygon, in the order in which they are connected</param>
+        /// <param name="algorithm">Which Line Algorithm to use</param>
+        /// <exception cref="ArgumentException">Must have 3 or more corners; Algorithm must produce ordered lines.</exception>
+        public PolygonArea(IEnumerable<Point> corners, Lines.Algorithm algorithm = Lines.Algorithm.DDA)
+        {
+            //corner initialization
+            if (corners is List<Point> crnrs)
+                _corners = crnrs;
+            else
+                _corners = corners.ToList();
```

Comment:
The check for `List<Point>` type is not an ideal implementation; namely because it does not make it obvious to a user that, if a list is passed, something different will occur.  This can actually affect a user in very subtle ways; as when a list is _not_ copied like this, any additions made _by the user_ to the list they passed in after the creation of the `PolygonArea` will also magically appear in the `_corners` list (because it's a reference to the same object); since `PolygonArea` does not allow this to occur, this can result in undefined behavior.

In my initial comment, I suggested just doing the `_corners = corners.ToList()` copy in this version, and providing another constructor with the following signature:
```CS
public PolygonArea(ref List<Point> corners, Lines.Algorithm algorithm = Lines.Algorithm.DDA)
{
    _corners = corners;
    // Do whatever else is needed
}
```

The use of `ref` here forces the caller to become aware that the reference is being stolen and they should refrain from using it after, because to use this constructor, the user must write the following code:
```CS
// Ref keyword MUST be included; if it is not, the IEnumerable version is called
var poly = new PolygonArea(ref myExistingList);
```

This forces the user to very explicitly state that they _do not_ want to copy the list, and accept any memory ramifications of doing so.